### ISSUE TITLE
[API Proposal]: default state for all objects

### ISSUE THREAD

**Issue by Xyncgas** [dotnet/runtime#64007]
### Background and motivation

We have the default keyword setting all bits to 0 (coming from a discussion where default / new() creates a default state for all instance was talked about : https://github.com/dotnet/csharplang/discussions/5666#discussioncomment-1999150)

but to get default state for all instance would require support from this side of .NET
Meanwhile the benefit of having a default state would mean all reference objects can now be something other than null, and mayhaps this can affect struct too, having a default state instead of a new instance saves memory where you have a lot default initialized objects.

Currently we can do it by ourselves using custom getter & setter / generic wrapper but I think it is worth considering adding this to .NET

### API Proposal

public class A{ }

var DefaultState = new A();
A DefaultState2 = default;

//These two variables should point to the same instance until they are accessed to be modifed,
For saving memory

### API Usage

Included above

### Alternative Designs

this is unique

### Risks

It will be transparent, if we only make it happen by assigning static default state to objects, and create a new instance when they are modified, unless we start changing what the ```default``` keyword means

**Comment by stephentoub** [comment]
> but to get default state for all instance would require support from this side of .NET

I'm not sure what API you're suggesting.  Are you asking for RuntimeHelpers.GetUninitializedObject?
https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.runtimehelpers.getuninitializedobject

**Comment by Frassle** [comment]
> mean all reference objects can now be something other than null

Not all reference types have a default value.
Not all reference types have a no parameters constructor (i.e. `new A()`).

> having a default state instead of a new instance saves memory where you have a lot default initialized objects.

Sure, that's why [flyweight](https://en.wikipedia.org/wiki/Flyweight_pattern) and [copy on write](https://en.wikipedia.org/wiki/Copy-on-write) patterns were invented. But the cost for the runtime to support those automatically for every type would be massive, both for engineering cost but also runtime performance those things aren't free.

**Comment by huoyaoyuan** [comment]
**The meaning of the word "default" is really, really unclear.** With DI container, a type can be configured to transient, scoped or singleton lifetime. It means that the "default" usable instance you will use may depend on other states. The complexity itself can not be exposed with a simple language keyword.

Global singleton pattern is simple, but it is anti-pattern sometimes, **and it still have variations**, for example lazy or eager creation, thread safety of creation etc.

**Comment by Xyncgas** [comment]
yes, I meant all objects has a default state that may or may not use the default constructor, but once the default state is defined in the code you can have all objects of this type to hold a reference to this static default state by default instead of null by default, that means ```A MyVariable;``` wouldn't be null if user had specified that A has a default state other than null.

Alternatively, this feature can allow compiler options to treat all objects with a default state regardless of what the state actually is -- that is instead of user defining what's a default state make all basic types non-nullable (removing null) and naturally a lot of things will not be null

**Comment by Xyncgas** [comment]
> **The meaning of the word "default" is really, really unclear.** With DI container, a type can be configured to transient, scoped or singleton lifetime. It means that the "default" usable instance you will use may depend on other states. The complexity itself can not be exposed with a simple language keyword.
> 
> Global singleton pattern is simple, but it is anti-pattern sometimes, **and it still have variations**, for example lazy or eager creation, thread safety of creation etc.

Well it's kinda what I said, but I am sorry I can't design the api myself, but I would like to add more ideas I will stay open minded and I am understanding it's something unlikely to happen

**Comment by Clockwork-Muse** [comment]
> make all basic types non-nullable (removing null) and naturally a lot of things will not be null

This isn't possible, for historical reasons.  That is, there's a lot of code that's of the form:
```csharp
if (staticvar == null)
{
    staticvar = InitializeStaticState();
}
```

**Comment by stephentoub** [comment]
Thank you for the interest and suggestions.  We're not going to make the proposed changes.

**Comment by huoyaoyuan** [comment]
> that is instead of user defining what's a default state make all basic types non-nullable (removing null) and naturally a lot of things will not be null

How?
References have to be initialized to something to be non-null. During the initialization, the default state is null. If the initializer is simple enough, the language has already provided a shortcut `= new()`.
**Initialization is very costful**, especially for arrays.

**Comment by Xyncgas** [comment]
> > that is instead of user defining what's a default state make all basic types non-nullable (removing null) and naturally a lot of things will not be null
> 
> How? References have to be initialized to something to be non-null. During the initialization, the default state is null. If the initializer is simple enough, the language has already provided a shortcut `= new()`. **Initialization is very costful**, especially for arrays.

Just assign a default state to objects for them to be non null, calling new() costs, assigning static default state isn't costly, so imagine an object gets a reference to a static state by default, but language design magics makes it instantiate when something modifies the variable / object

**Comment by Clockwork-Muse** [comment]
> Just assign a default state to objects for them to be non null, calling new() costs, assigning static default state isn't costly, so imagine an object gets a reference to a static state by default, but language design magics makes it instantiate when something modifies the variable / object (that's default state)

... you can't get out of calling `new()` at all, and the compiler already assigns a static state (essentially, blits in all 0s).  So copy-on-write tracking is an additional, rather expensive, cost.

**Comment by Xyncgas** [comment]
> > Just assign a default state to objects for them to be non null, calling new() costs, assigning static default state isn't costly, so imagine an object gets a reference to a static state by default, but language design magics makes it instantiate when something modifies the variable / object (that's default state)
> 
> ... you can't get out of calling `new()` at all, and the compiler already assigns a static state (essentially, blits in all 0s). So copy-on-write tracking is an additional, rather expensive, cost.

I meant copy-on-write for the first time, and having a default state actually saves memory. Imagine you have a list that represents 100 floors in a hotel, of course each floor starts with a default state of number of clients, instead of having 100 instance just initialize a default static state and assign it as a reference to all the floors, then if someone decides to get a room at 2nd floor the rt now instantiate the object.

**Comment by Xyncgas** [comment]
> the compiler already assigns a static state (essentially, blits in all 0s)
The behavior you are talking about is too low in the abstraction

**Comment by Clockwork-Muse** [comment]
> I meant copy-on-write for the first time

Yes, I know.  The problem is that now the runtime has to keep track of "has the first mutate happened", and **that** is the really expensive part.  Especially as most objects are expected to be mutated "immediately", which means you're going to incur the initialization costs regardless.

> Imagine you have a list that represents 100 floors in a hotel

This is a poor example, because relying on the list index for the floor number is bad practice, which means you actually _should_ have 100 different floor objects with an id of some sort (or more realistically, a collection of rooms with a floor id), which means you have no "default state" to rely on.  

This isn't to say that there aren't uses for copy-on-write, but usually such objects require more targeted knowledge of a domain or anticipated use - it's not something that works well as a blanket application.

> The behavior you are talking about is too low in the abstraction at least in default states things can be something other than 0 

And this is what a constructor is for - assigning default state.  Most of the cost of a constructor is not the initialization of member variables, but other parts of object initialization.

> (the default state can be decided at rt not just in ct too, but it's a default state for all variables uninitialized, essentially varaible declared without assigning anything)

Being able to have separate default states at run time and compile time would be a major design flaw, because it means a programmer wouldn't be able to know what the base state of their object was.  Any default state would have to be a compile time feature... which we already have, as part of object initialization.

**Comment by Xyncgas** [comment]
> because it means a programmer wouldn't be able to know what the base state of their object was. Any default state would have to be a compile time feature...

The programmer can decide the default state I didn't say they can't or shouldn't, but the reason I said you can decide it at rt is because static objects are initialized at rt, say if you assign something static to an object at rt the static object that's assigning is created before the object you are trying to assign it's created before any instances of objects

I also didn't say you have to define the default state, if it's not defined in ct then the default state will be something DEFAULT as is something that should be considered in the proposal's api so the specifications for what's the default state for native .NET types are (0 for blittable, String.Empty for string, etc)

> And this is what a constructor is for
Calling the constructor isn't the answer to what I am proposing, because it solves not having creating an instance, because when you call the constructor you are calling new, that allocates memory, but having a default state doesn't allocate memory, that's a big difference

> This is a poor example

it doesn't matter, it's an example, you can't possibly know all the use case and say it's useless. Meanwhile I've listed reasoning for why it's useful, it doesn't matter you think it's a bad example or not, the memories are still being wasted just by initializing a bunch of floors in this case (most of them are empty) and it has proven itself possibilities to waste memory, so I would say the example did a good job on explaining how having a default state for objects can save memory.

> which means you actually should have 100 different floor objects with an id

like I said it doesn't matter, I like to have a list with 100 different floors as array of array, some how through profiling I find that it makes my application faster by consisdent 3% (for example) also the place I am working at use no-sql so there isn't columns for room-id floor-id that's just how you do things but we don't think that's the future

> The problem is that now the runtime has to keep track of "has the first mutate happened", and that is the really expensive part.

It's not, your regular GC already keeps tracks of a bunch of pointers that's intptr size, you instantiating objects are already expensive on the allocating memory part, whether you assign a created instance or a static object to a variable it's linear complexity, the object mutation is also linear complexity for the way users mutating it (using with keyword on records for example) in contrast of the rt doing it, meanwhile the system can be designed in a way, ok, you are already keeping track of the static objects, now just watch anyone that tries to change them, you are not letting them to (because they are immutable default states) and instead create a new instance which the rt is not going to worry about whether someone is trying to change them because it's not watching them.

> Especially as most objects are expected to be mutated "immediately"

It depends buddy.

> which means you're going to incur the initialization costs regardless.

Bingo

So basically it has the benefit of reducing memory costs initially, it kinda instantiate themselves lazily, .NET rt isn't going to compact & GC collect memory often so if it ate something it is not going to spit out for a while so if the memory is limited and your browser is using a bunch memory you and the user would kinda wanna see it use lower memory but meanwhile having a default state means you don't see null it means you don't pay the computation efforts upfront it means I can let people choose if they want to opt-in for this feature

**Comment by Clockwork-Muse** [comment]
> but having a default state doesn't allocate memory, that's a big difference

That doesn't negate anything I said, though.  Presumably, instead of the default initialization value being `null` (historically, literally 0`), it would be a reference to a pre-initialized value somewhere in memory.  The problem is that now the runtime has to have a way to distinguish whether that reference points to one of those pre-initialized values (which it absolutely can't mutate) or a "normal" reference (which it can).  This is a **large, difficult** problem with a relatively small payoff, especially when in most cases a non-null default value isn't going to be valid any more than `null` is.

> instead create a new instance which the rt is not going to worry about whether someone is trying to change them because it's not watching them.

There's two problems with this:
 1) The check for whether it's a default object has to go _somewhere_, and in most proposals I've seen it needs to be performed on **every** object access.
 2) The runtime currently doesn't really have a concept of object/reference purity, which means it currently has no way to know whether any given operation is going to need to replace the reference or not.

> it doesn't matter, it's an example, you can't possibly know all the use case and say it's useless.

You're right, I can't.  If you're advocating for such a feature, though, you need to be able to come up with a use-case that the team finds minimally reasonable.

The actual problem with your use-case is that it's too "primitive" - you haven't abstracted your domain sufficiently, and doing so is likely to render the need for this as a language feature moot.  For instance, instead of directly exposing a list of floors with default values, exposing an index operator (or method) that returns a newly-created default object that you can safely mutate.

> So basically...

This is implying that there's some flaw or oversight in your application design, that you're allocating references that you don't need to, or not sufficiently considering when a value might be "not present".  Having the value be "some default" doesn't necessarily magically fix these problems, and in most cases good application design will solve the issue the same way.

> also the place I am working at use no-sql so there isn't columns for room-id floor-id that's just how you do things but we don't think that's the future

🤨 
This isn't an SQL-vs-noSQL problem, and having an id for the floor doesn't mean that there's a floor object somewhere (I was more thinking of just the floor number itself).  Even then, it is occasionally useful for noSQL systems to have cross-reference ids for various reasons, even if only internally (ie, someone making a query doesn't see them, but the database itself combines documents when data is requested).
